utilities/hydrofabric_test/test_ngen.py

```diff
@@ -0,0 +1,198 @@
+        raise NgenRunException
+
+def make_realization(workdir: Path, config: Formulation, forcing: Forcing, time: Time, include_routing: bool=False):
+    """Generates a ngen realization file based on the user defined formulation.
+
+    Args:
+        workdir (Path): Location to stage the realization file in.
+        config (Formulation): Formulation to build.
+        forcing (Forcing): Forcing configuration to use.
+        time (Time): Simulation time range.
+        include_routing (bool, optional): Whether or not to add a routing block to the realization. Defaults to False.
+                                          When true, assumes the routing config is in the workdir and called test_routing_config.yaml
+    """
+
+    f = Formulation(name=config['name'], params=config)
+    r = Realization(formulations=[f], forcing=forcing)
+    if(include_routing):
+        routing = {'t_route_config_file_with_path': workdir/'test_routing_config.yaml'}
+    else:
+        routing = None
+    g = NgenRealization(global_config=r, time=time, routing=routing)
+    with open(workdir/"realization.json", 'w') as fp:
+         fp.write( g.json(by_alias=True, exclude_none=True, indent=4))
+
+#TODO on failue of check_call, output log to console???
+def test_ngen(hydrofabric: Union[str, Path], config, tag=""):
+    """_summary_
+
+    Args:
+        hydrofabric (Union[str, Path]): path to geopackage hydrofabric to test
+        config (General): the user configuration
+        tag (str, optional): A tag to prefix the temporary directory the model runs in. Defaults to "".
+
+    Raises:
+        NgenRunException: _description_
+    """
+    ngen = config.ngen
+    workdir = config.workdir
+    tmp_dir = Path(mkdtemp(prefix=tag, dir=workdir))
+    chdir(tmp_dir)
+    make_geojson(hydrofabric)
+    #This is required if we want to test routing embedded in ngen, but right now it gets tested stand-alone
+    stage_routing_config(hydrofabric, tmp_dir)
+    make_realization(tmp_dir, config.formulation, config.forcing, config.time, config.routing.embedded)
+    cmd = f'{ngen} catchments.geojson all nexus.geojson all realization.json'
+    try:
+        with open('ngen.log', 'w') as log_file:
+            subprocess.check_call(cmd, stdout=log_file, stderr=log_file, shell=True, cwd=tmp_dir)
+        print(f"ngen run successful on {hydrofabric}")
+    except:
+        print(f"Running ngen on hydrofabric {hydrofabric} failed, check {tmp_dir}/ngen.log")
+        raise NgenRunException
+    test_routing(tmp_dir, hydrofabric)
+    chdir(workdir)
+
+def test_routing(workdir, hydrofabric):
+    """Test routing via ngen_main in an external call.
+
+    Args:
+        workdir (Path): working directory to run the test in
+        hydrofabric (Path): which hydrofabric to test
+    """
+
+    if not (workdir/'test_routing_config.yaml').exists():
+        stage_routing_config(hydrofabric, workdir)
+    cmd = 'python -m ngen_routing.ngen_main -f test_routing_config.yaml'
```

Comment:
@mattw-nws, is this still correct?

Reply:
Only for legacy t-route versions (i.e. t-route's ngen branch). For their master branch, it's 
```
python -m nwm_routing -V4 -f /path/to/routing_config.yaml
```
... I even [documented it!](/NOAA-OWP/ngen/blob/master/doc/PYTHON_ROUTING.md#running-t-route-separately-with-ngen-output)

Reply:
This PR predates that by several months (Nov 2022 vs Jun 2023).  Nonetheless, is in need of updating if we want to test the main branch...will make some adjustments to this soon.